evm_loader/program/src/executor.rs

```diff
@@ -574,15 +574,17 @@ impl<'a, B: AccountStorage> Machine<'a, B> {
                 },
                 _ => {
                     self.executor.state.exit_commit().map_err(|e| (Vec::new(), ExitReason::from(e)))?;
-                    self.executor.state.set_code(address, return_value);
+                    self.executor.state.set_code(address, &return_value);
                 }
             };
         }
 
-        let runtime = match self.runtime.last_mut() {
-            Some((runtime, _)) => runtime,
-            None => return Err((Vec::new(), reason))
+        let runtime = match (self.runtime.last_mut(), &reason) {
+            (Some((last_runtime, _)), _) => last_runtime,
+            (None, ExitReason::Revert(_)) => return Err((return_value, reason)),
+            (None, _) => return Err((Vec::<u8>::new(), reason))
         };
```

Comment:
I think something like this will look better. And won't waste memory.
```rust
let return_value = if reason.is_succeed() { 
     ...
    Vec::new() 
} else { 
    return_value
};

let runtime = match self.runtime.last_mut() {
    Some((runtime, _)) => runtime,
    None => return Err((return_value, reason))
};
```

Reply:
### Memory

If i do it in suggested way it **will** waste more memory because of program flow
Have a look please at my code structure. If **`last runtime`** actually exists it won't waste the memory at all and vise versa it'll need to return existent `return_value` or create empty `Vec`

From the other viewpoint if I agree with you and the **`last_runtime`** doesn't exist i'll have to at least create an empty Vec if it's not Revert

### Structure 

I my honest opinion suggested structure is better because of Logic doesn't get spread, gets isolated

Reply:
The biggest reason I don't like it is the large number of parenthesizes. It make code hard to read and understand.
You also allocating memory for additional copy of `return_value` when you don't need it.